keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+  (unless there is a good namespace to put it in that would work for all
+  clusters).
+  - To always have the default configuration cluster scoped, policy defintions
+    could be allowed to reference two configuration CRDs, one for cluster scope
+    and another for namespace scope. Having two CRDs is VERY undesirable.
+
+None of these options seem satisfactory to me.
+
+#### Access to namespace metadata
+
+- Namespace labels and annotations are the most commonly needed fields not
+  already available in the resource being validated. Note that
+  namespaceSelectors already allow matches to examine namespace levels, but we
+  also have use cases that need to be able to inspects the fields in CEL
+  expressions.
+
+#### Transition rules
+
+- Will provide access to "object" and "oldObject" in CEL expressions. These will
+  be the same as in AdmissionReview.
+- On CREATE, "oldObject" will be null.
+- On DELETE, "object" will be null.
+
+If we add "CEL expression scoping" (see above section), we will also need to
+consider how scoped fields are handled for create/update/delete. Note that CRD
+validation rules have transition rules which are only evaluated when both "self"
+and "oldSelf" are present. 
+
+#### Composition utilities
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+cel-policy-template "terms" are very convenient. But we don't have consensus on
+if they are sufficiently essential. Need to either agree on something or drop
+this.
+<<[/UNRESOLVED]>>
+
+- Ability to define a sub-expression and then use it in multiple validation
+  rules (xref cel-policy-template "terms")
+
+#### Safety Features
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+Decide exactly what we are willing to commit to here.
+<<[/UNRESOLVED]>>
+
+- Configurable admission blocking write requests made internally in
+  kube-apiserver during server startup (like RBAC default policy reconciliation)
+  making it impossible for a server to start up healthy. (This is not specific to CEL?)
+- Ability to skip specific resource types - Admission Controller Webhook
+  configuration rule cannot exclude specific resources:
+  https://github.com/kubernetes/kubernetes/issues/92157
+
+#### Aggregated API servers
+
+Main complications (provided by @liggitt):
+
+- The API server validating/persisting the ValidatingAdmissionPolicy
+  instances isn't the same one serving aggregated types, so wouldn't necessarily
+  have schema info to check type safety.
+
+- The aggregated API server is responsible for enforcing admission on its custom
+  types, so the implementation that reads ValidatingAdmissionPolicy
+  instances and enforces them would have to live in k8s.io/apiserver and be
+  active in aggregated API servers to enforce admission on aggregated types
+  effectively (same as admission webhooks today).
+
+Plan:
+
+- Do not offer type checking for aggregated types.
+- Support ValidatingAdmissionPolicy in aggregated API servers.
+
+#### Metrics
+
+Goals:
```

Comment:
At what granularity do you need latency metrics?

Reply:
I'm in favor of a metric per {policy, validation expression}

The next level of granularity would be {policy, configuration, validation expression}, but that the number of configurations can become quite large, so I'm hesitant to go that far.

I've put this in the KEP.

Reply:
Are we concerned that policy/expression names in cleartext/unacld metrics might leak policy surface?